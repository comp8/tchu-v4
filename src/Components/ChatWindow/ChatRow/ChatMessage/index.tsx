import React from "react";
import Emoticon from "../../../Emoticon";

function Text({ str }: { str: string }) {
  return <span className='chat-message__text'>{str}</span>
}

interface ChatMessageProps {
  message: string;
  emotes: { [id: string]: string[] };
}

export default function ChatMessage(props: ChatMessageProps) {
  const { message, emotes: emotesObj } = props;

  let elements: React.ReactElement[] = [];

  if (emotesObj) {
    let emotesArr: [string, number, number][] = [];
    Object.keys(emotesObj).forEach(key => {
      [...emotesObj[key]].forEach(position => {
        let match = /^(\d+)-(\d+)$/.exec(position);
        if (match) {
          let a = parseInt(match[1]);
          let b = parseInt(match[2]);
          emotesArr.push([key, a, b]);
        }
      });
    });
    if (emotesArr.length > 0) {
      emotesArr.sort((a, b) => a[1] - b[1]);

      let i = 0;
      for (const emote of emotesArr) {
        const emoteId = emote[0];
        const begin = emote[1];
        const end = emote[2];

        if (i < begin) {
          elements.push(<Text str={message.slice(i, begin)} />);
        }
        elements.push(<Emoticon id={emoteId} />);
        i = end + 1;
      }
      if (i < message.length) {
        elements.push(<Text str={message.slice(i)} />);
      }
    }
  }

  if (elements.length <= 0) {
    elements.push(<Text str={message} />);
  }

  return (
    <span className='chat-message'>
      {
        ...elements
      }
    </span>
  );
}
