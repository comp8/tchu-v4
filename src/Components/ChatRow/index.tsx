import React from "react";
import { ChatItem } from "../../common/Types";
import { arbitraryColor } from "../../common/utils";
import ChatMessage from "../ChatWindow/ChatRow/ChatMessage";
import UserName from "../UserName";

interface ChatRowProps {
  data: ChatItem;
  className?:string;
}

export default function ChatRow(props: ChatRowProps) {

  const { data } = props;

  return (
    <li className={props.className} data-channel={data.channel}>
      <UserName
        className='chat-name-wrapper'
        badges={data.userstate.badges}
        dname={data.userstate["display-name"]}
        uname={data.userstate.username}
        color={data.userstate.color || arbitraryColor(data.userstate["user-id"])}
        delimiter
      />
      <ChatMessage
        message={data.message}
        emotes={data.userstate.emotes}
      />
    </li>
  );
}
