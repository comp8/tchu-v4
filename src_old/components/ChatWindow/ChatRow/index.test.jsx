import React from 'react';
import shallowRenderer from 'react-test-renderer/shallow';

import ChatRow from '.';

describe('', () => {
  test('', () => {
    const shallow = shallowRenderer.createRenderer();
    shallow.render(<ChatRow data={{ channel: 'test_channel_123', message: 'hello world', userstate: {} }} />);
    let result = shallow.getRenderOutput();
    expect(result).toMatchSnapshot();
    expect(result.type).toBe('li');
  });
  test('', () => {
    const shallow = shallowRenderer.createRenderer();
    shallow.render(<ChatRow data={{ channel: 'test_channel_456', message: 'hello, world!', userstate: { color: '#112233' } }} />);
    let result = shallow.getRenderOutput();
    expect(result).toMatchSnapshot();
    expect(result.props.children[0].props.color).toEqual('#112233');
  });
  test('', () => {
    const shallow = shallowRenderer.createRenderer();
    shallow.render(<ChatRow data={{ channel: 'test_channel_777', message: 'qwerty', userstate: { "user-id": '1234' } }} />);
    let result = shallow.getRenderOutput();
    expect(result).toMatchSnapshot();
    expect(result.props.children[0].props.color).not.toEqual('#112233');
  });
  test('', () => {
    const shallow = shallowRenderer.createRenderer();
    shallow.render(<ChatRow data={{ channel: 'test_channel_456', message: 'hello, world!', userstate: { color: '#112233', "user-id": '1234' } }} />);
    let result = shallow.getRenderOutput();
    expect(result).toMatchSnapshot();
    expect(result.props.children[0].props.color).toEqual('#112233');
  });
})