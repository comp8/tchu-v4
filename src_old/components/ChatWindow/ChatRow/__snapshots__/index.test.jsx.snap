// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`  1`] = `
<li
  data-channel="test_channel_123"
>
  <UserName
    className="chat-name-wrapper"
    color="#0000ff"
    delimiter={true}
  />
  <ChatMessage
    message="hello world"
  />
</li>
`;

exports[`  2`] = `
<li
  data-channel="test_channel_456"
>
  <UserName
    className="chat-name-wrapper"
    color="#112233"
    delimiter={true}
  />
  <ChatMessage
    message="hello, world!"
  />
</li>
`;

exports[`  3`] = `
<li
  data-channel="test_channel_777"
>
  <UserName
    className="chat-name-wrapper"
    color="#00797b"
    delimiter={true}
  />
  <ChatMessage
    message="qwerty"
  />
</li>
`;

exports[`  4`] = `
<li
  data-channel="test_channel_456"
>
  <UserName
    className="chat-name-wrapper"
    color="#112233"
    delimiter={true}
  />
  <ChatMessage
    message="hello, world!"
  />
</li>
`;
